includes/data-stores/class-wc-admin-reports-orders-data-store.php

```diff
@@ -233,320 +149,136 @@ public function get_data( $query_args ) {
+			if ( $forbidden_statuses ) {
+				$subqueries[] = "status NOT IN ( '" . implode( "','", $forbidden_statuses ) . "' )";
+			}
 		}
 
-		self::update( $order );
-	}
-
-	/**
-	 * Syncs order information when a refund is deleted.
-	 *
-	 * @param int $refund_id Refund ID.
-	 * @param int $order_id Order ID.
-	 */
-	public static function sync_on_refund_delete( $refund_id, $order_id ) {
-		self::sync_order( $order_id );
+		return implode( " $operator ", $subqueries );
 	}
 
 	/**
-	 * Update the database with stats data.
+	 * Enriches the order data.
 	 *
-	 * @param WC_Order $order Order to update row for.
-	 * @return int|bool|null Number or rows modified or false on failure.
+	 * @param array $orders_data Orders data.
+	 * @param array $query_args  Query parameters.
 	 */
-	public static function update( $order ) {
-		global $wpdb;
-		$table_name = $wpdb->prefix . self::TABLE_NAME;
-
-		if ( ! $order->get_id() || ! $order->get_date_created() ) {
-			return false;
-		}
-
-		if ( ! in_array( $order->get_status(), parent::get_report_order_statuses(), true ) ) {
-			$wpdb->delete(
-				$table_name,
-				array(
-					'order_id' => $order->get_id(),
-				)
-			);
-			return;
-		}
-
-		$data   = array(
-			'order_id'           => $order->get_id(),
-			'date_created'       => $order->get_date_created()->date( 'Y-m-d H:i:s' ),
-			'num_items_sold'     => self::get_num_items_sold( $order ),
-			'gross_total'        => $order->get_total(),
-			'coupon_total'       => $order->get_total_discount(),
-			'refund_total'       => $order->get_total_refunded(),
-			'tax_total'          => $order->get_total_tax(),
-			'shipping_total'     => $order->get_shipping_total(),
-			'net_total'          => (float) $order->get_total() - (float) $order->get_total_tax() - (float) $order->get_shipping_total(),
-			'returning_customer' => self::is_returning_customer( $order ),
-			'status'             => self::normalize_order_status( $order->get_status() ),
-		);
-		$format = array(
-			'%d',
-			'%s',
-			'%d',
-			'%f',
-			'%f',
-			'%f',
-			'%f',
-			'%f',
-			'%f',
-			'%d',
-			'%s',
-		);
-
-		// Ensure we're associating this order with a Customer in the lookup table.
-		$order_user_id        = $order->get_customer_id();
-		$customers_data_store = new WC_Admin_Reports_Customers_Data_Store();
-
-		if ( 0 === $order_user_id ) {
-			$email = $order->get_billing_email( 'edit' );
-
-			if ( $email ) {
-				$customer_id = $customers_data_store->get_or_create_guest_customer_from_order( $order );
-
-				if ( $customer_id ) {
-					$data['customer_id'] = $customer_id;
-					$format[] = '%d';
+	protected function include_extended_info( &$orders_data, $query_args ) {
+		foreach ( $orders_data as $key => $order_data ) {
+			$order                       = wc_get_order( $order_data['order_id'] );
+			$extended_info               = new ArrayObject();
+			$extended_info['products']   = array();
+			$extended_info['categories'] = array();
+			foreach ( $order->get_items() as $item ) {
```

Comment:
I think we had this discussion before on another PR ( with @psealock me thinks ), and I'm fine with doing this in a follow-up, but couldn't we be potentially duplicating some queries here for getting taxonomies?

Reply:
The only idea I had for the categories again was to maybe pre-fetch all the categories prior to the foreach loops once, and create a map of them $terms[ id ] => $term->name sort of thing. Or perhaps even build up that mapping and one for the products as we loop through the orders.

Quite possibly I'm being overly paranoid here, and mysql or WP query caching will already help us out here.

Reply:
Took a stab at optimizing this a bit in https://github.com/woocommerce/wc-admin/pull/1246/commits/4c6119fd4f2d204acfb4fc6b3cc1b1223c5dd640 but it may need a bit more work.

Feedback is welcome 😄